sprint_8-solution-time-and-duration

Maks1-15 opened the pull request.

Maks1-15 committed:
8 фз
Добавил интерфейс для валидации данных
Добавил дополнительную логику в FileBackedTaskManager
Добавил тесты на все классы
Добавлена функциональность связанная со временем задач
Добавлена логика обработки задач с временем

Maks1-15 committed:
8 фз
Добавил дополнительную логику в FileBackedTaskManager
Добавил тесты на все классы
Добавлена функциональность связанная со временем задач
Добавлена логика обработки задач с временем

Umka7389 commented on src/Main.java:
```diff
@@ -1,49 +1,35 @@
-
-        FileBackedTaskManager FileBackedTaskManager = (FileBackedTaskManager) Managers.getFileBackedTaskManager(file);
+        // изменить file
+        // добавить все все тесты
```
Не стоит оставлять рабочие комментарии в коде

Umka7389 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -44,13 +90,15 @@ public void removeTasks() {
     }
 
+    // В этом методе как-то не догадался как именно через Stream. И не подскажешь, тут же может возникнуть исключение
+    // т.к. мы меняем коллекцию. Чтобы избежать этого, я должен создать новую коллекцию и с ней проводить манипуляции или как?
```
Нет, новую коллекцию создавать не надо, надо просто правильно обработать текущую, возможно ConcurrentMidificationException, но я не могу предусмотреть все возможные при который может возникнуть исключение). Обработка в простом цикле вроде бы позволяет избежать этого исключения

Umka7389 commented on src/service/InMemoryTaskManager.java:
```diff
@@ -89,21 +137,30 @@ public Subtask getByIdSubtask(int id) {
 
+    // Еще вопрос, мы должны работать с объектами напрямую, или брать их копии?
+    // ситуация в том, что любую задачу мы можем изменить путем использования settera, не прибегая к методу update,
+    // как я понимаю, это не доработка и это нужно исправить путем создания копии задачи и работы с ней, так?
```
Нет, не надо работать с копиями, в данном случае вам это не нужно, если есть сомнения по сеттерам вы можете его всегда закрыть или вообще не писать)

Umka7389 commented on test/service/InMemoryHistoryManagerTest.java:
```diff
@@ -21,38 +23,42 @@ public void beforeEach() {
+        assertEquals(1, hm.addTaskInMapHistory(task1));
     }
 
+//    @Test
```
Не оставляете закомментированный код

Umka7389 reviewed (commented):
Молодец, хорошая работа! Замечания в комментариях

Maks1-15 committed:
8 фз
Исправлены ошибки
Добавлено клонирование задач при создании для избежания неявного изменения задач через сеттеры
Добавлены соответствующие тесты
Изменена структура некоторых методов

Umka7389 reviewed (commented):
Молодец, хорошая работа!
